File: app/services/task_service.py
from collections.abc import Sequence
from datetime import UTC, datetime
from pathlib import Path
from uuid import uuid4

from app.core.config import settings
from app.core.time_utils import to_naive_utc
from app.models.enums import TaskStatus
from app.models.task import Label, Task, TaskAttachment
from app.repositories.task_repo import TaskRepository
from app.repositories.user_repo import UserRepository
from app.schemas.label import LabelCreate
from app.schemas.task import TaskAttachmentCreate, TaskCreate, TaskUpdate
from app.services.exceptions import BadRequestError, ConflictError, NotFoundError
from app.services.experience import ExperienceAwarder


def _serialize_user(user) -> dict | None:
    if user is None:
        return None
    return {
        "id": user.id,
        "username": user.username,
        "avatar_url": user.avatar_url,
        "position": user.position,
        "member_status": user.member_status,
        "github_login": user.github_login,
        "is_placeholder": user.is_placeholder,
    }


def _serialize_label(label: Label) -> dict:
    return {
        "id": label.id,
        "name": label.name,
        "color": label.color,
        "created_at": label.created_at,
        "updated_at": label.updated_at,
    }


def _serialize_related_task(task: Task) -> dict:
    return {
        "id": task.id,
        "title": task.title,
        "status": task.status,
    }


def _serialize_attachment(attachment: TaskAttachment) -> dict:
    return {
        "id": attachment.id,
        "name": attachment.name,
        "url": attachment.url,
        "created_by": _serialize_user(attachment.created_by),
        "created_at": attachment.created_at,
    }


class TaskService:
    def _validate_status_transition(
        self,
        old_status: TaskStatus,
        new_status: TaskStatus,
    ) -> None:
        if old_status == new_status:
            return

        if new_status == TaskStatus.done and old_status != TaskStatus.review:
            raise BadRequestError("Task can be moved to done only from review")

    def _should_clear_approval(self, old_status: TaskStatus, new_status: TaskStatus) -> bool:
        if old_status == new_status:
            return False
        if new_status == TaskStatus.review:
            return True
        return old_status in {TaskStatus.review, TaskStatus.done} and new_status != TaskStatus.done

    def __init__(
        self,
        task_repo: TaskRepository,
        user_repo: UserRepository,
        experience_awarder: ExperienceAwarder,
    ):
        self.task_repo = task_repo
        self.user_repo = user_repo
        self.experience_awarder = experience_awarder

    def serialize_task_list_item(self, task: Task) -> dict:
        labels = [_serialize_label(task_label.label) for task_label in task.labels]
        related_count = len(task.left_relations) + len(task.right_relations)
        return {
            "id": task.id,
            "title": task.title,
            "status": task.status,
            "difficulty": task.difficulty,
            "deadline": task.deadline,
            "creator": _serialize_user(task.creator),
            "assignee": _serialize_user(task.assignee),
            "labels": labels,
            "attachments_count": len(task.attachments),
            "related_tasks_count": related_count,
            "github_issue_number": task.github_issue_number,
            "github_url": (
                f"https://github.com/{settings.github_repo}/issues/{task.github_issue_number}"
                if task.github_issue_number is not None
                else None
            ),
            "created_at": task.created_at,
            "updated_at": task.updated_at,
        }

    def serialize_task_detail(self, task: Task) -> dict:
        data = self.serialize_task_list_item(task)
        related_tasks = [relation.right_task for relation in task.left_relations] + [
            relation.left_task for relation in task.right_relations
        ]
        data.update(
            {
                "description": task.description,
                "attachments": [
                    _serialize_attachment(attachment) for attachment in task.attachments
                ],
                "related_tasks": [_serialize_related_task(item) for item in related_tasks],
                "creator_id": task.creator_id,
                "assignee_id": task.assignee_id,
                "approved_by_id": task.approved_by_id,
                "approved_at": task.approved_at,
            }
        )
        return data

    def _validate_deadline(self, deadline: datetime | None) -> datetime | None:
        """
        Проверяет, что deadline в будущем.
        Возвращает deadline без timezone для совместимости с БД.
        """
        if deadline is None:
            return None
        if deadline.tzinfo is None:
            deadline = deadline.replace(tzinfo=UTC)
        if deadline <= datetime.now(UTC):
            raise BadRequestError("Deadline must be in the future")
        return to_naive_utc(deadline)

    async def _ensure_user_exists(self, user_id: int | None) -> None:
        """Проверяет существование пользователя."""
        if user_id is None:
            return
        user = await self.user_repo.get_user_by_id(user_id)
        if not user:
            raise NotFoundError(f"User with id {user_id} not found")

    async def create_task(self, data: TaskCreate, creator_id: int) -> dict:
        """Создаёт новую задачу."""
        clean_deadline = self._validate_deadline(data.deadline)

        await self._ensure_user_exists(creator_id)
        await self._ensure_user_exists(data.assignee_id)

        fields = data.model_dump()
        fields["deadline"] = clean_deadline

        task = await self.task_repo.create_task(fields, creator_id=creator_id)
        return self.serialize_task_detail(await self.get_task_model_by_id(task.id))

    async def get_tasks(
        self,
        status: TaskStatus | None = None,
        assignee_id: int | None = None,
        difficulty: int | None = None,
        limit: int = 100,
        offset: int = 0,
    ) -> Sequence[dict]:
        """Получает список задач с фильтрами."""
        tasks = await self.task_repo.get_tasks(status, assignee_id, difficulty, limit, offset)
        return [self.serialize_task_list_item(task) for task in tasks]

    async def get_task_model_by_id(self, task_id: int) -> Task:
        """Получает задачу по ID."""
        task = await self.task_repo.get_task_by_id(task_id)
        if not task:
            raise NotFoundError(f"Task with id {task_id} not found")
        return task

    async def get_task_by_id(self, task_id: int) -> dict:
        return self.serialize_task_detail(await self.get_task_model_by_id(task_id))

    async def update_task(self, task_id: int, data: TaskUpdate) -> dict:
        """Обновляет обычные поля задачи (не статус и не исполнителя)."""
        task = await self.get_task_model_by_id(task_id)
        update_data = data.model_dump(exclude_unset=True)

        if "deadline" in update_data:
            update_data["deadline"] = self._validate_deadline(update_data["deadline"])

        task = await self.task_repo.update_task(task, update_data)
        return self.serialize_task_detail(await self.get_task_model_by_id(task.id))

    async def change_status(self, task_id: int, new_status: TaskStatus) -> dict:
        """
        Меняет статус задачи.

        Если переход non-done -> done, вызывает хук начисления опыта.
        Атомарно: если award_for_task упадёт, статус не изменится.
        """
        task = await self.get_task_model_by_id(task_id)
        old_status = task.status
        self._validate_status_transition(old_status, new_status)

        moving_to_done_from_review = (
            old_status == TaskStatus.review and new_status == TaskStatus.done
        )
        if moving_to_done_from_review and not task.approved_at:
            raise BadRequestError("Task must be approved before moving to done")

        task = await self.task_repo.set_status(
            task,
            new_status,
            clear_approval=self._should_clear_approval(old_status, new_status),
        )

        if old_status != TaskStatus.done and new_status == TaskStatus.done:
            await self.experience_awarder.award_for_task(task)

        return self.serialize_task_detail(await self.get_task_model_by_id(task.id))

    async def assign_task(self, task_id: int, assignee_id: int | None) -> dict:
        """
        Назначает или снимает исполнителя задачи.

        Разрешено даже для задач в статусе done (MVP).
        XP не переносится и не отзывается.
        """
        task = await self.get_task_model_by_id(task_id)
        await self._ensure_user_exists(assignee_id)
        task = await self.task_repo.set_assignee(task, assignee_id)
        return self.serialize_task_detail(await self.get_task_model_by_id(task.id))

    async def approve_task(self, task_id: int, approver_id: int) -> dict:
        task = await self.get_task_model_by_id(task_id)
        await self._ensure_user_exists(approver_id)

        if task.status != TaskStatus.review:
            raise BadRequestError("Only tasks in review can be approved")

        approved_at = datetime.now(UTC).replace(tzinfo=None)
        task = await self.task_repo.approve_task(task, approver_id, approved_at)
        return self.serialize_task_detail(await self.get_task_model_by_id(task.id))

    async def list_labels(self) -> list[Label]:
        return await self.task_repo.list_labels()

    async def create_label(self, data: LabelCreate) -> Label:
        existing = await self.task_repo.get_label_by_name_ci(data.name)
        if existing is not None:
            raise ConflictError("Label name already exists")
        return await self.task_repo.create_label(data.name, data.color)

    async def set_task_labels(self, task_id: int, label_ids: list[int]) -> dict:
        await self.get_task_model_by_id(task_id)
        labels = await self.task_repo.get_labels_by_ids(label_ids)
        if len(labels) != len(label_ids):
            raise NotFoundError("One or more labels were not found")
        await self.task_repo.replace_task_labels(task_id, label_ids)
        return self.serialize_task_detail(await self.get_task_model_by_id(task_id))

    async def list_attachments(self, task_id: int) -> list[dict]:
        await self.get_task_model_by_id(task_id)
        attachments = await self.task_repo.list_attachments(task_id)
        return [_serialize_attachment(attachment) for attachment in attachments]

    async def create_attachment(
        self,
        task_id: int,
        data: TaskAttachmentCreate,
        created_by_id: int,
    ) -> dict:
        await self.get_task_model_by_id(task_id)
        await self._ensure_user_exists(created_by_id)
        attachment = await self.task_repo.create_attachment(
            task_id,
            data.name,
            data.url,
            created_by_id,
        )
        return _serialize_attachment(attachment)

    async def create_uploaded_attachment(
        self,
        task_id: int,
        filename: str,
        content: bytes,
        created_by_id: int,
    ) -> dict:
        await self.get_task_model_by_id(task_id)
        await self._ensure_user_exists(created_by_id)
        if not content:
            raise BadRequestError("Uploaded file is empty")

        original_name = Path(filename).name or "attachment"
        safe_name = "".join(
            char if char.isalnum() or char in {".", "-", "_"} else "-" for char in original_name
        ).strip(".-")
        if not safe_name:
            safe_name = "attachment"

        upload_dir = Path(settings.upload_dir) / "task-attachments"
        upload_dir.mkdir(parents=True, exist_ok=True)
        stored_name = f"{uuid4().hex}-{safe_name}"
        file_path = upload_dir / stored_name
        file_path.write_bytes(content)

        attachment = await self.task_repo.create_attachment(
            task_id,
            original_name,
            f"/uploads/task-attachments/{stored_name}",
            created_by_id,
        )
        return _serialize_attachment(attachment)

    async def delete_attachment(self, task_id: int, attachment_id: int) -> None:
        await self.get_task_model_by_id(task_id)
        attachment = await self.task_repo.get_attachment(task_id, attachment_id)
        if attachment is None:
            raise NotFoundError("Attachment not found")
        await self.task_repo.delete_attachment(attachment)

    async def list_related_tasks(self, task_id: int) -> list[dict]:
        await self.get_task_model_by_id(task_id)
        tasks = await self.task_repo.list_related_tasks(task_id)
        return [_serialize_related_task(task) for task in tasks]

    async def set_related_tasks(self, task_id: int, task_ids: list[int]) -> list[dict]:
        await self.get_task_model_by_id(task_id)
        if task_id in task_ids:
            raise BadRequestError("Task cannot be related to itself")
        related_tasks = await self.task_repo.get_tasks_by_ids(task_ids)
        if len(related_tasks) != len(task_ids):
            raise NotFoundError("One or more related tasks were not found")
        await self.task_repo.replace_related_tasks(task_id, task_ids)
        return await self.list_related_tasks(task_id)

File: backend/src/main/java/com/skilltracker/service/TaskService.java
package com.skilltracker.service;

import com.skilltracker.domain.Label;
import com.skilltracker.domain.Task;
import com.skilltracker.domain.TaskAttachment;
import com.skilltracker.domain.TaskLabel;
import com.skilltracker.domain.TaskRelation;
import com.skilltracker.domain.TaskStatus;
import com.skilltracker.dto.LabelCreateRequest;
import com.skilltracker.dto.LabelResponse;
import com.skilltracker.dto.RelatedTaskResponse;
import com.skilltracker.dto.TaskAttachmentCreateRequest;
import com.skilltracker.dto.TaskAttachmentResponse;
import com.skilltracker.dto.TaskCreateRequest;
import com.skilltracker.dto.TaskDetailResponse;
import com.skilltracker.dto.TaskListItemResponse;
import com.skilltracker.dto.TaskUpdateRequest;
import com.skilltracker.exception.BadRequestException;
import com.skilltracker.exception.ConflictException;
import com.skilltracker.exception.NotFoundException;
import com.skilltracker.repository.LabelRepository;
import com.skilltracker.repository.OffsetLimit;
import com.skilltracker.repository.TaskAttachmentRepository;
import com.skilltracker.repository.TaskLabelRepository;
import com.skilltracker.repository.TaskRelationRepository;
import com.skilltracker.repository.TaskRepository;
import com.skilltracker.repository.TaskSpecifications;
import com.skilltracker.repository.UserRepository;
import java.time.Clock;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class TaskService {

    private final TaskRepository taskRepository;
    private final UserRepository userRepository;
    private final LabelRepository labelRepository;
    private final TaskLabelRepository taskLabelRepository;
    private final TaskAttachmentRepository taskAttachmentRepository;
    private final TaskRelationRepository taskRelationRepository;
    private final ExperienceAwarder experienceAwarder;
    private final AttachmentStorage attachmentStorage;
    private final TaskMapper taskMapper;
    private final ReadBack readBack;
    private final Clock clock;

    public TaskService(
            TaskRepository taskRepository,
            UserRepository userRepository,
            LabelRepository labelRepository,
            TaskLabelRepository taskLabelRepository,
            TaskAttachmentRepository taskAttachmentRepository,
            TaskRelationRepository taskRelationRepository,
            ExperienceAwarder experienceAwarder,
            AttachmentStorage attachmentStorage,
            TaskMapper taskMapper,
            ReadBack readBack,
            Clock clock) {
        this.taskRepository = taskRepository;
        this.userRepository = userRepository;
        this.labelRepository = labelRepository;
        this.taskLabelRepository = taskLabelRepository;
        this.taskAttachmentRepository = taskAttachmentRepository;
        this.taskRelationRepository = taskRelationRepository;
        this.experienceAwarder = experienceAwarder;
        this.attachmentStorage = attachmentStorage;
        this.taskMapper = taskMapper;
        this.readBack = readBack;
        this.clock = clock;
    }

    @Transactional(readOnly = true)
    public List<TaskListItemResponse> getTasks(
            TaskStatus status, Integer assigneeId, Integer difficulty, int limit, int offset) {
        return taskMapper.toListItems(taskRepository
                .findAll(
                        TaskSpecifications.matching(status, assigneeId, difficulty),
                        OffsetLimit.of(limit, offset, Sort.by(Sort.Direction.DESC, "id")))
                .getContent());
    }

    @Transactional(readOnly = true)
    public TaskDetailResponse getTask(Integer taskId) {
        return taskMapper.toDetail(requireTask(taskId));
    }

    @Transactional
    public TaskDetailResponse createTask(TaskCreateRequest request, Integer creatorId) {
        LocalDateTime deadline = validateDeadline(request.deadline());
        requireUserExists(creatorId);
        requireUserExists(request.assigneeId());

        Task task = new Task(request.title(), creatorId);
        task.setDescription(request.description());
        task.setDifficulty(request.difficulty());
        task.setDeadline(deadline);
        task.setAssigneeId(request.assigneeId());

        taskRepository.save(task);
        readBack.flushAndDetach();
        return taskMapper.toDetail(requireTask(task.getId()));
    }

    @Transactional
    public TaskDetailResponse updateTask(Integer taskId, TaskUpdateRequest request) {
        Task task = requireTask(taskId);

        if (request.title().isPresent() && request.title().value() != null) {
            task.setTitle(request.title().value());
        }
        if (request.description().isPresent()) {
            task.setDescription(request.description().value());
        }
        if (request.difficulty().isPresent() && request.difficulty().value() != null) {
            task.setDifficulty(request.difficulty().value());
        }
        if (request.deadline().isPresent()) {
            task.setDeadline(validateDeadline(request.deadline().value()));
        }

        taskRepository.save(task);
        readBack.flushAndDetach();
        return taskMapper.toDetail(requireTask(taskId));
    }

    /**
     * Moves a task through the workflow and, on the first transition into {@code done}, pays out its
     * XP in the same transaction. The task row is locked so two concurrent completions cannot both
     * observe a non-done status and award twice.
     */
    @Transactional
    public TaskDetailResponse changeStatus(Integer taskId, TaskStatus newStatus) {
        Task locked = taskRepository
                .findByIdForUpdate(taskId)
                .orElseThrow(() -> new NotFoundException("Task with id " + taskId + " not found"));

        TaskStatus oldStatus = locked.getStatus();
        validateStatusTransition(oldStatus, newStatus);

        if (oldStatus == TaskStatus.REVIEW && newStatus == TaskStatus.DONE && locked.getApprovedAt() == null) {
            throw new BadRequestException("Task must be approved before moving to done");
        }

        locked.setStatus(newStatus);
        if (shouldClearApproval(oldStatus, newStatus)) {
            locked.clearApproval();
        }
        taskRepository.saveAndFlush(locked);

        if (oldStatus != TaskStatus.DONE && newStatus == TaskStatus.DONE) {
            experienceAwarder.awardForTask(locked);
        }

        readBack.flushAndDetach();
        return taskMapper.toDetail(requireTask(taskId));
    }

    @Transactional
    public TaskDetailResponse assignTask(Integer taskId, Integer assigneeId) {
        Task task = requireTask(taskId);
        requireUserExists(assigneeId);

        task.setAssigneeId(assigneeId);
        taskRepository.save(task);
        readBack.flushAndDetach();
        return taskMapper.toDetail(requireTask(taskId));
    }

    @Transactional
    public TaskDetailResponse approveTask(Integer taskId, Integer approverId) {
        Task task = requireTask(taskId);
        requireUserExists(approverId);

        if (task.getStatus() != TaskStatus.REVIEW) {
            throw new BadRequestException("Only tasks in review can be approved");
        }

        task.approve(approverId, LocalDateTime.now(clock));
        taskRepository.save(task);
        readBack.flushAndDetach();
        return taskMapper.toDetail(requireTask(taskId));
    }

    @Transactional(readOnly = true)
    public List<LabelResponse> listLabels() {
        return labelRepository.findAllByOrderByNameAsc().stream()
                .map(LabelResponse::from)
                .toList();
    }

    @Transactional
    public LabelResponse createLabel(LabelCreateRequest request) {
        if (labelRepository.findByNameIgnoreCase(request.name()).isPresent()) {
            throw new ConflictException("Label name already exists");
        }
        return LabelResponse.from(labelRepository.save(new Label(request.name(), request.color())));
    }

    @Transactional
    public TaskDetailResponse setTaskLabels(Integer taskId, List<Integer> labelIds) {
        requireTask(taskId);

        List<Label> labels = labelIds.isEmpty() ? List.of() : labelRepository.findByIdIn(labelIds);
        if (labels.size() != labelIds.size()) {
            throw new NotFoundException("One or more labels were not found");
        }

        taskLabelRepository.deleteAllByTaskId(taskId);
        for (Integer labelId : labelIds) {
            taskLabelRepository.save(new TaskLabel(taskId, labelId));
        }
        readBack.flushAndDetach();

        return taskMapper.toDetail(requireTask(taskId));
    }

    @Transactional(readOnly = true)
    public List<TaskAttachmentResponse> listAttachments(Integer taskId) {
        requireTask(taskId);
        return taskMapper.attachments(taskId);
    }

    @Transactional
    public TaskAttachmentResponse createAttachment(
            Integer taskId, TaskAttachmentCreateRequest request, Integer createdById) {
        requireTask(taskId);
        requireUserExists(createdById);

        TaskAttachment attachment =
                taskAttachmentRepository.save(new TaskAttachment(taskId, request.name(), request.url(), createdById));
        readBack.flushAndDetach();
        return taskMapper.toAttachment(
                taskAttachmentRepository.findByIdWithAuthor(attachment.getId()).orElseThrow());
    }

    @Transactional
    public TaskAttachmentResponse createUploadedAttachment(
            Integer taskId, String filename, byte[] content, Integer createdById) {
        requireTask(taskId);
        requireUserExists(createdById);
        if (content == null || content.length == 0) {
            throw new BadRequestException("Uploaded file is empty");
        }

        String displayName = attachmentStorage.displayName(filename);
        String url = attachmentStorage.store(displayName, content);

        TaskAttachment attachment =
                taskAttachmentRepository.save(new TaskAttachment(taskId, displayName, url, createdById));
        readBack.flushAndDetach();
        return taskMapper.toAttachment(
                taskAttachmentRepository.findByIdWithAuthor(attachment.getId()).orElseThrow());
    }

    @Transactional
    public void deleteAttachment(Integer taskId, Integer attachmentId) {
        requireTask(taskId);
        TaskAttachment attachment = taskAttachmentRepository
                .findByTaskIdAndId(taskId, attachmentId)
                .orElseThrow(() -> new NotFoundException("Attachment not found"));
        taskAttachmentRepository.delete(attachment);
    }

    @Transactional(readOnly = true)
    public List<RelatedTaskResponse> listRelatedTasks(Integer taskId) {
        requireTask(taskId);
        return taskMapper.relatedTasks(taskId);
    }

    @Transactional
    public List<RelatedTaskResponse> setRelatedTasks(Integer taskId, List<Integer> relatedTaskIds) {
        requireTask(taskId);
        if (relatedTaskIds.contains(taskId)) {
            throw new BadRequestException("Task cannot be related to itself");
        }

        Set<Integer> existing =
                relatedTaskIds.isEmpty() ? Set.of() : new HashSet<>(taskRepository.findExistingIds(relatedTaskIds));
        if (existing.size() != relatedTaskIds.size()) {
            throw new NotFoundException("One or more related tasks were not found");
        }

        taskRelationRepository.deleteInvolving(taskId);
        for (Integer relatedTaskId : relatedTaskIds) {
            taskRelationRepository.save(TaskRelation.between(taskId, relatedTaskId));
        }
        readBack.flushAndDetach();

        return taskMapper.relatedTasks(taskId);
    }

    private void validateStatusTransition(TaskStatus oldStatus, TaskStatus newStatus) {
        if (oldStatus == newStatus) {
            return;
        }
        if (newStatus == TaskStatus.DONE && oldStatus != TaskStatus.REVIEW) {
            throw new BadRequestException("Task can be moved to done only from review");
        }
    }

    private boolean shouldClearApproval(TaskStatus oldStatus, TaskStatus newStatus) {
        if (oldStatus == newStatus) {
            return false;
        }
        if (newStatus == TaskStatus.REVIEW) {
            return true;
        }
        return (oldStatus == TaskStatus.REVIEW || oldStatus == TaskStatus.DONE) && newStatus != TaskStatus.DONE;
    }

    /** Deadlines are stored as naive UTC and must lie in the future. */
    private LocalDateTime validateDeadline(LocalDateTime deadline) {
        if (deadline == null) {
            return null;
        }
        if (!deadline.isAfter(LocalDateTime.now(clock))) {
            throw new BadRequestException("Deadline must be in the future");
        }
        return deadline;
    }

    private Task requireTask(Integer taskId) {
        return taskRepository
                .findById(taskId)
                .orElseThrow(() -> new NotFoundException("Task with id " + taskId + " not found"));
    }

    private void requireUserExists(Integer userId) {
        if (userId != null && !userRepository.existsById(userId)) {
            throw new NotFoundException("User with id " + userId + " not found");
        }
    }
}
